jax/api_util.py

```diff
@@ -82,6 +85,49 @@ def argnums_partial(f, dyn_argnums, args):
   dyn_args = tuple(args[i] for i in dyn_argnums)
   return _argnums_partial(f, dyn_argnums, fixed_args), dyn_args
 
+def donation_vector(donate_argnums, args, kwargs) -> Tuple[bool, ...]:
+  """Returns a tuple with a boolean value for each leaf in args."""
+  res = []
+  for i, arg in enumerate(args):
+    donate = bool(i in donate_argnums)
+    res.extend((donate,) * tree_structure(arg).num_leaves)
+  res.extend((False,) * tree_structure(kwargs).num_leaves)
+  return tuple(res)
+
+def rebase_donate_argnums(donate_argnums, static_argnums) -> Tuple[int, ...]:
+  """Shifts donate to account for static.
+
+  >>> rebase_donate_argnums((3, 4), (0, 1))
+  (1, 2)
+
+  Args:
+    donate_argnums: An iterable of ints.
+    static_argnums: An iterable of ints.
+
+  Returns:
+    A tuple of unique, sorted integer values based on donate_argnums with each
+    element offset to account for static_argnums.
+  """
+  if not (static_argnums or donate_argnums):
+    return tuple(sorted(donate_argnums))
+
+  static_argnums = sorted(set(static_argnums))
+  donate_argnums = sorted(set(donate_argnums))
+  i = j = o = 0
+  out = []
+  while j < len(donate_argnums):
```

Comment:
This is gonna be super slow, but it's only invoked at compile time so it doesn't really matter.